warehouse/integrations/github/utils.py

```diff
@@ -0,0 +1,355 @@
+                raise GitHubPublicKeyMetaAPIError(
+                    f"Key is not a dict but: {public_key}",
+                    "public_key_api.format_error",
+                )
+
+            attributes = set(public_key)
+            if not expected_attributes <= attributes:
+                raise GitHubPublicKeyMetaAPIError(
+                    "Missing attribute in key: "
+                    f"{sorted(expected_attributes - attributes)}",
+                    "public_key_api.format_error",
+                )
+
+            result.append(
+                {"key": public_key["key"], "key_id": public_key["key_identifier"]}
+            )
+
+        self.public_keys_cache = public_keys
+        return result
+
+    def _check_public_key(self, github_public_keys, key_id):
+        for record in github_public_keys:
+            if record["key_id"] == key_id:
+                return record["key"]
+
+        raise InvalidTokenLeakRequest(
+            f"Key {key_id} not found in github public keys", reason="wrong_key_id"
+        )
+
+    def _check_signature(self, payload, public_key, signature):
+        try:
+            loaded_public_key = serialization.load_pem_public_key(
+                data=public_key.encode("utf-8"), backend=default_backend()
+            )
+            loaded_public_key.verify(
+                signature=base64.b64decode(signature),
+                data=payload.encode("utf-8"),
+                # This validates the ECDSA and SHA256 part
+                signature_algorithm=ECDSA(algorithm=SHA256()),
+            )
+        except InvalidSignature as exc:
+            raise InvalidTokenLeakRequest(
+                "Invalid signature", "invalid_signature"
+            ) from exc
+        except Exception as exc:
+            # Maybe the key is not a valid ECDSA key, maybe the data is not properly
+            # padded, etc. So many things can go wrong...
+            raise InvalidTokenLeakRequest(
+                "Invalid cryptographic values", "invalid_crypto"
+            ) from exc
+
+
+def _analyze_disclosure(request, disclosure_record, origin):
+
+    metrics = request.find_service(IMetricsService, context=None)
+
+    metrics.increment(f"warehouse.token_leak.{origin}.recieved")
+
+    try:
+        disclosure = TokenLeakDisclosureRequest.from_api_record(
+            record=disclosure_record
+        )
+    except InvalidTokenLeakRequest as exc:
+        metrics.increment(f"warehouse.token_leak.{origin}.error.{exc.reason}")
+        return
+
+    macaroon_service = request.find_service(IMacaroonService, context=None)
+    try:
+        database_macaroon = macaroon_service.find_from_raw(
+            raw_macaroon=disclosure.token
+        )
+    except InvalidMacaroon:
+        metrics.increment(f"warehouse.token_leak.{origin}.error.invalid")
+        return
+
+    metrics.increment(f"warehouse.token_leak.{origin}.valid")
+
+    macaroon_service.delete_macaroon(macaroon_id=str(database_macaroon.id))
```

Comment:
Maybe instead of outright deleting the macaroon, we need a way to mark it as "disabled"?

Reply:
Ah I was just thinking about something like this
<img width="805" alt="Capture d’écran 2021-03-16 à 19 28 51" src="https://user-images.githubusercontent.com/1457576/111361196-dfe88380-868d-11eb-8499-c6895518161c.png">

Reply:
Yes, I definitely think we should store an event for this as part of this PR. Up to you whether you want to include some "disabling" functionality here or save that for a followup PR. I think it's OK if we just remove these for now, the priority should be getting this integration set up.

Reply:
For the record, here's a screenshot of the result:
<img width="824" alt="Capture d’écran 2021-03-16 à 23 42 57" src="https://user-images.githubusercontent.com/1457576/111393396-fbb54f00-86b8-11eb-9f9f-37740ee7032b.png">

Reply:
I've added "automatically" in the title